File: Leetcode 1743. Restore the Array From Adjacent Pairs/lc 1743.java
class Solution {
    public int[] restoreArray(int[][] adjacentPairs) {
        ArrayList<Integer> ans=new ArrayList<>();
        HashMap<Integer,ArrayList<Integer>> graph=new HashMap<>();
        
        for(int i=0;i<adjacentPairs.length;i++){
            int f=adjacentPairs[i][0];
            int s=adjacentPairs[i][1];

            if(graph.containsKey(f)){
                ArrayList<Integer> adj=graph.get(f);
                adj.add(s);
                graph.put(f,adj);
            }else{
                ArrayList<Integer> adj=new ArrayList<>();
                adj.add(s);
                graph.put(f,adj);
            }

            if(graph.containsKey(s)){
                ArrayList<Integer> adj=graph.get(s);
                adj.add(f);
                graph.put(s,adj);
            }else{
                ArrayList<Integer> adj=new ArrayList<>();
                adj.add(f);
                graph.put(s,adj);
            }
            
        }
        
        for(Map.Entry<Integer,ArrayList<Integer>> m:graph.entrySet()){
            ArrayList<Integer> val=m.getValue();
            if(val.size()==1){
                ans.add(m.getKey());
                ans.add(val.get(0));
                break;
            }
        }

        while(ans.size()<graph.size()){
            int lastElement=ans.get(ans.size()-1);
            int secondLastElement=ans.get(ans.size()-2);
            if(graph.get(lastElement).size()==1){
                ans.add(graph.get(lastElement).get(0));
            }else{
                int f=graph.get(lastElement).get(0);
                if(f!=secondLastElement){
                    ans.add(f);
                }else{
                    ans.add(graph.get(lastElement).get(1));
                }
            }
        }
        int[] res=new int[ans.size()];
        for(int i=0;i<ans.size();i++){
            res[i]=ans.get(i);
        }
        return res;
    }
}

File: Leetcode 1743. Restore the Array From Adjacent Pairs/lc 1743.cpp
class Solution {
public:
    vector<int> restoreArray(vector<vector<int>>& adjacentPairs) {
        vector<int> ans;
        unordered_map<int,vector<int>> graph;

        for(auto pair:adjacentPairs){
            graph[pair[0]].emplace_back(pair[1]);
            graph[pair[1]].emplace_back(pair[0]);
        }

        for(auto pair:graph){
            if(pair.second.size()==1){
                ans.emplace_back(pair.first);
                ans.emplace_back(pair.second[0]);
                break;
            }
        }

        while(ans.size()<graph.size()){
            int lastEle=ans[ans.size()-1];
            int secondLast=ans[ans.size()-2];
            vector<int> pair=graph[lastEle];
            if(pair.size()==1){
                ans.emplace_back(graph[lastEle][0]);
                
            }else{
                if(pair[0]!=secondLast){
                    ans.emplace_back(pair[0]);
                }else{
                    ans.emplace_back(pair[1]);
                }
            }
        }
        return ans;
    }
};